cmd/thermal-recorder/snapshot.go

```diff
@@ -0,0 +1,57 @@
+package main
+
+import (
+	"errors"
+	"image"
+	"image/color"
+	"image/png"
+	"math"
+	"os"
+	"path"
+
+	"github.com/TheCacophonyProject/lepton3"
+)
+
+var (
+	previousSnapshotID = 0
+	snapshotDir        string
+)
+
+func newSnapshot() error {
+	f := frameLoop.Previous()
+	if f == nil {
+		return errors.New("no frames yet")
+	}
+	g16 := image.NewGray16(image.Rect(0, 0, lepton3.FrameCols, lepton3.FrameRows))
+	// Max and min are needed for normalization of the frame
+	var valMax uint16
+	var valMin uint16 = math.MaxUint16
+	var id int
+	for _, row := range f {
+		for _, val := range row {
+			id += int(val)
+			valMax = uint16(math.Max(float64(valMax), float64(val)))
+			valMin = uint16(math.Min(float64(valMin), float64(val)))
+		}
+	}
+
+	// Check if frame had already been processed
+	if id == previousSnapshotID {
+		return nil
+	}
+	previousSnapshotID = id
```

Comment:
A few things here:
* as it stands newSnapshot could end up being called concurrently which means that access to previousSnapshotID needs to be synchronised (with a Mutex or similar)
* if rate limiting is applied in the dbus call then you can be sure that newSnapshot will be getting called slow enough that every call will be dealing with a new frame. This would mean there's no need to check if the frame has changed.

I'm happy to look at this if you like, after the PR has landed.